### PR TITLE
[turf, junction] new ports

### LINKED ISSUE #17433: [New Port Request] junction
Library name: junction

Library description:
A high performance multi-producer, multi-consumer, lock-free, thread-safe hash map.
See benchmarks [here](https://preshing.com/20160201/new-concurrent-hash-maps-for-cpp/). 

Source repository URL: https://github.com/preshing/junction

Anything else that is useful to know when adding (such as optional features the library may have that should be included):
It's a popular library, with 1.1k stars on GitHub. It performs better than most of the concurrent hash map libraries currently included in vcpkg ports.

### PR DESCRIPTION
**Pull request by mathisloge**
**Describe the pull request**
adds turf and junction
- #### What does your PR fix?  
  Fixes #17433

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  linux, windows static

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

### CODE REVIEW

**Comment by mathisloge** [comment]
@xanderdunn  please review and if you have time, please test

**Comment by xanderdunn** [comment]
@mathisloge huge thanks for putting this together. After further development with junction we've decided to move away from it for several reasons:
- junction only supports raw pointers as keys and values in the hash map. smart pointers are not supported. This puts the burden of memory management on the developer.
- linking with junction's turf dependency can cause problems when trying to link with `-fsanitize=address` because both turf and the address sanitizer library define the same symbols.
- Requires calling an update function from every thread that's accessing the concurrent hash map or memory will leak
We've replaced our usage with libcuckoo, which is already a part of vcpkg, and so far it appears to solve the above issues

@mathisloge your original hesitation around the junction package was in the right direction. 

junction can still be used successfully if a developer is careful, I think it's your call whether to include the new port. I'm happy with closing my original issue either way.

**Comment by mathisloge** [comment]
thanks for the explanations @xanderdunn . 

I would move the decision if this should be included in vcpkg to the person who reviews this PR. 
But as both seems to not actively developed it might be a good solution to move turf and junction into a registry. I can move them into my https://github.com/mathisloge/vcpkg-registry

**Comment by mathisloge** [comment]
@JackBoosY please decide on the comment above. Thanks

File: ports/turf/export-module.patch [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
++)
++
++install(EXPORT turfTargets
++        FILE turfTargets.cmake
++        NAMESPACE turf::
++        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/turf
++)
++
++include(CMakePackageConfigHelpers)
++
```
Comment (JackBoosY) [inline ports/turf/export-module.patch:48]:
```suggestion
+file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/Config.cmake.in
+[[
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/junctionTargets.cmake")
+
+check_required_components(junction)
+]]
+)
```

File: ports/turf/export-module.patch [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
++configure_package_config_file(${CMAKE_CURRENT_SOURCE_DIR}/Config.cmake.in
```
Comment (JackBoosY) [inline ports/turf/export-module.patch:49]:
```suggestion
+configure_package_config_file(${CMAKE_CURRENT_BINARY_DIR}/Config.cmake.in

```

File: ports/junction/Config.cmake.in [hunk 3]
```diff
@@ -0,0 +1,5 @@
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/junctionTargets.cmake")
+
+check_required_components(junction)
```
Comment (JackBoosY) [inline ports/junction/Config.cmake.in:5]: We don't need this file, see above.

File: ports/junction/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/Macros.cmake DESTINATION ${SOURCE_PATH}/cmake)
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DJUNCTION_WITH_SAMPLES=OFF
+        -DTURF_PREFER_CPP11=1
+)
+
+vcpkg_install_cmake()
```
Comment (JackBoosY) [inline ports/junction/portfile.cmake:21]:
```suggestion
vcpkg_install_cmake()
vcpkg_copy_pdbs()
```

File: ports/junction/use-vcpkg-turf.patch [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
++        NAMESPACE junction::
++        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/junction
++)
++
++include(CMakePackageConfigHelpers)
++
++configure_package_config_file(${CMAKE_CURRENT_SOURCE_DIR}/Config.cmake.in
++    "${CMAKE_CURRENT_BINARY_DIR}/junctionConfig.cmake"
++    INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/junction
++)
```
Comment (JackBoosY) [inline ports/junction/use-vcpkg-turf.patch:61]: Same as above.

File: ports/turf/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DTURF_PREFER_CPP11=ON
+        -DTURF_MAKE_INSTALLABLE=ON
+)
+
+vcpkg_install_cmake()
```
Comment (JackBoosY) [inline ports/turf/portfile.cmake:21]:
```suggestion
vcpkg_install_cmake()
vcpkg_copy_pdbs()
```

File: ports/turf/vcpkg.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "turf",
+  "version-string": "2017-01-13",
```
Comment (JackBoosY) [inline ports/turf/vcpkg.json:3]:
```suggestion
  "version-date": "2017-01-13",
```

File: ports/junction/vcpkg.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "junction",
+  "version-string": "2018-02-18",
```
Comment (JackBoosY) [inline ports/junction/vcpkg.json:3]:
```suggestion
  "version-date": "2018-02-18",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/turf/vcpkg.json (hunk 7 above)
Comment (mathisloge) [reply ports/turf/vcpkg.json]: resolved https://github.com/microsoft/vcpkg/pull/17540/commits/99596a7d4a40273c129ae9291f1f2d9712bb5bbd

File: ports/junction/vcpkg.json (hunk 8 above)
Comment (mathisloge) [reply ports/junction/vcpkg.json]: resolved https://github.com/microsoft/vcpkg/pull/17540/commits/99596a7d4a40273c129ae9291f1f2d9712bb5bbd

File: ports/junction/Config.cmake.in (hunk 3 above)
Comment (mathisloge) [reply ports/junction/Config.cmake.in]: resolved https://github.com/microsoft/vcpkg/pull/17540/commits/99596a7d4a40273c129ae9291f1f2d9712bb5bbd

File: ports/turf/export-module.patch (hunk 2 above)
Comment (mathisloge) [reply ports/turf/export-module.patch]: resolved https://github.com/microsoft/vcpkg/pull/17540/commits/99596a7d4a40273c129ae9291f1f2d9712bb5bbd

File: ports/turf/export-module.patch (hunk 1 above)
Comment (mathisloge) [reply ports/turf/export-module.patch]: resolved https://github.com/microsoft/vcpkg/pull/17540/commits/99596a7d4a40273c129ae9291f1f2d9712bb5bbd

**Comment by mathisloge** [comment]
@JackBoosY  whats your opinion on https://github.com/microsoft/vcpkg/pull/17540#issuecomment-827883450

**Comment by JackBoosY** [comment]
@mathisloge In my opinion, if the port is used by many users, it can be added to vcpkg even if there are some defects, unless the upstream has abandoned it.
vcpkg provides methods to build, install and use ports, but if the port itself contains bugs, we should let the upstream solve them.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Waiting for #17528 merge.

**Comment by mathisloge** [comment]
I really don't know if the library is used by many users. @xanderdunn was the first one requesting this lib, but then switched back to another library. turf (last commit 4 years ago) and junction(last commit 3 years ago) are both experimental libraries. And from the comments in some issues from the upstream maintainer it seems like it wouldn't be further developed.

For this reason, I don't know if these two are usable as vcpkg ports. But this decision is up to you. :)

**Comment by JackBoosY** [comment]
turf has almost 500 stars and junction has over 1000 stars, so I think they can be added in vcpkg.

**Comment by mathisloge** [comment]
@JackBoosY merged master after #17528 was merged and now CI is good

File: ports/junction/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO preshing/junction
+    REF 5ad3be7ce1d3f16b9f7ed6065bbfeacd2d629a08 # 2018-02-18
+    SHA512 a3daee41e45be0ad79a4bdb2745dee04f1418020df5914dd04692304890936c9da60840efee249a2cc0618eea572102ca4b7ecc6f6711a37dac461c6b894fba1
+    PATCHES 
+        "use-vcpkg-turf.patch"
+)
+
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/Macros.cmake DESTINATION ${SOURCE_PATH}/cmake)
```
[the remark this answers is not in the document]
Comment (mathisloge) [reply ports/junction/portfile.cmake:11]: will change later today. thanks for the review.

**Comment by strega-nil-ms** [comment]
Marking `requires:discussion`, as I want to seriously consider whether we wish to include two experimental and unmaintained libraries inside vcpkg; I think that @mathisloge is right, and that registries are likely the correct way forward for these ports.

**Comment by strega-nil-ms** [comment]
Since there is no upstream interest in maintaining these libraries, the builtin vcpkg registry is not the right place for them. @mathisloge, the perfect place for these is in your custom registry 😄 

Thank you for the PR!

**Comment by mathisloge** [comment]
thanks for the feedback  @strega-nil-ms  have added them to https://github.com/mathisloge/vcpkg-registry

**Comment by mathisloge** [comment]
maybe #17433 should be closed as well? @strega-nil-ms

**Pull request closed without merging**